fix segfault and slowdown when broker exceeds max number of open files

grondo opened the pull request:
This PR fixes the segv reported in issue #3505 and adds the reproducer described in that issue.

I have this marked as WIP because another issue remains. Every 10-20 times I run the reproducer, instead of failing immediately, the failed jobs appear to hang for an indeterminate amount of time, and the flux-broker process spins at 100% CPU. It is a challenge figuring out exactly what is happening during that time though, and after many seconds, the problem resolves itself.

I hope to find at fix that issue in this PR, but I don't get anywhere soon we may want to merge this one as is.

grondo commented:
> I have this marked as WIP because another issue remains.

I understand this problem now.

When the flux-broker exceeds the max number of open files, there is a good chance that the `accept(2)` call will fail in `connector-local` when, for example a new job shell attempts to connect to the local broker. When `server_accept()` in librouter fails to accept a new connection, it just returns immediately and the module goes back to sleep in the reactor. However, since the `accept()` failed, the socket is immediately ready and the `accept()` is tried again, fails, etc, until something else in the broker frees a file descriptor and the `accept()` can succeed.

In some cases, the resulting spin loop slows down other work sufficiently such that the broker is stuck in this condition for some time (several seconds). Meanwhile the problem is exacerbated by the fact that the broker main thread is flooded with keepalive messages (one before entering the reactor, and one just after waking up), making things even slower.

We can see the effect by looking at the keepalive counters for `connector-local` between a run of the reproducer that avoids the condition:

```
{
 "#request (tx)": 141,
 "#request (rx)": 10,
 "#response (tx)": 9,
 "#response (rx)": 98,
 "#event (tx)": 0,
 "#event (rx)": 0,
 "#keepalive (tx)": 692,
 "#keepalive (rx)": 0
}
```

and one that hits the condition:

```
{
 "#request (tx)": 141,
 "#request (rx)": 10,
 "#response (tx)": 9,
 "#response (rx)": 99,
 "#event (tx)": 0,
 "#event (rx)": 0,
 "#keepalive (tx)": 1450686,
 "#keepalive (rx)": 0
}
```

One mitigation that seems to work is to "pause" handling usock server connection attempts for a short period of time whenever `accept(2)` returns `EMFILE` or `ENFILE`. This eliminates the excessive load of keepalive requests on the broker (down to say 100s per second instead of millions per second), and allows other parts of the system to proceed and (hopefully) close fds so that the next `accept()` attempt may succeed.

It is also quite simple.

Another approach might be to hold some fds in reserve as described in [this SO post](https://stackoverflow.com/questions/47179793/how-to-gracefully-handle-accept-giving-emfile-and-close-the-connection), however that probably isn't going to work out in the flux-broker since it is a threaded process, and something else may grab our reserve fds before accept() can use them.

Therefore, at this point I'd lean toward something like this:

```diff
diff --git a/src/common/librouter/usock.c b/src/common/librouter/usock.c
index e1d241c49..c94a1113e 100644
--- a/src/common/librouter/usock.c
+++ b/src/common/librouter/usock.c
@@ -467,6 +467,16 @@ static struct usock_conn *server_accept (struct usock_server *server,
     return conn;
 }
 
+static void timeout_cb (flux_reactor_t *r,
+                        flux_watcher_t *w,
+                        int revents,
+                        void *arg)
+{
+    struct usock_server *server = arg;
+    flux_watcher_start (server->w);
+    flux_watcher_destroy (w);
+    fprintf (stderr, "Starting usock server\n");
+}
 
 static void server_cb (flux_reactor_t *r,
                        flux_watcher_t *w,
@@ -478,8 +488,25 @@ static void server_cb (flux_reactor_t *r,
     if ((revents & FLUX_POLLIN)) {
         struct usock_conn *conn;
 
-        if (!(conn = server_accept (server, r)))
+        if (!(conn = server_accept (server, r))) {
+            if ((errno == ENFILE || errno == EMFILE)
+                && !getenv ("USOCK_DISABLE_EMFILE_WORKAROUND")) {
+                /*  Too many open files. Do not just back to sleep in the
+                 *   reactor since we'll wake right back up again. Instead
+                 *   disable this callback until after a short pause, giving
+                 *   time for fds to be closed hopefully...
+                 */
+                flux_watcher_t *tw = flux_timer_watcher_create (r,
+                                                                0.01,
+                                                                0,
+                                                                timeout_cb,
+                                                                server);
+                flux_watcher_start (tw);
+                flux_watcher_stop (w);
+                fprintf (stderr, "Too many open files. Pausing usock server\n");
+            }
             return;
+        }
         if (zlist_append (server->connections, conn) < 0) {
             usock_conn_destroy (conn);
             return;

```

I left some debug statements in the patch above to satisfy myself that the w/a is activated:

```
2021-02-14T20:29:41.381041Z broker.err[0]: socketpair: Too many open files
2021-02-14T20:29:41.381946Z broker.err[0]: socketpair: Too many open files
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
Starting usock server
Too many open files. Pausing usock server
2021-02-14T20:29:41.478584Z job-exec.err[0]: exec_start_cmd failed: Too many open files
Starting usock server
Too many open files. Pausing usock server
2021-02-14T20:29:41.488407Z job-exec.err[0]: exec_start_cmd failed: Too many open files
Starting usock server
```

With that patch applied, I cannot reproduce the "hang" anymore, unless setting `USOCK_DISABLE_EMFILE_WORKAROUND=t` in the environment of the test.

garlick commented:
Nice sleuthing!  The workaround sounds good to me!

On Sun, Feb 14, 2021, 12:33 PM Mark Grondona <notifications@github.com>
wrote:

> I have this marked as WIP because another issue remains.
>
> I understand this problem now.
>
> When the flux-broker exceeds the max number of open files, there is a good
> chance that the accept(2) call will fail in connector-local when, for
> example a new job shell attempts to connect to the local broker. When
> server_accept() in librouter fails to accept a new connection, it just
> returns immediately and the module goes back to sleep in the reactor.
> However, since the accept() failed, the socket is immediately ready and
> the accept() is tried again, fails, etc, until something else in the
> broker frees a file descriptor and the accept() can succeed.
>
> In some cases, the resulting spin loop slows down other work sufficiently
> such that the broker is stuck in this condition for some time (several
> seconds). Meanwhile the problem is exacerbated by the fact that the broker
> main thread is flooded with keepalive messages (one before entering the
> reactor, and one just after waking up), making things even slower.
>
> We can see the effect by looking at the keepalive counters for
> connector-local between a run of the reproducer that avoids the condition:
>
> {
>  "#request (tx)": 141,
>  "#request (rx)": 10,
>  "#response (tx)": 9,
>  "#response (rx)": 98,
>  "#event (tx)": 0,
>  "#event (rx)": 0,
>  "#keepalive (tx)": 692,
>  "#keepalive (rx)": 0
> }
>
> and one that hits the condition:
>
> {
>  "#request (tx)": 141,
>  "#request (rx)": 10,
>  "#response (tx)": 9,
>  "#response (rx)": 99,
>  "#event (tx)": 0,
>  "#event (rx)": 0,
>  "#keepalive (tx)": 1450686,
>  "#keepalive (rx)": 0
> }
>
> One mitigation that seems to work is to "pause" handling usock server
> connection attempts for a short period of time whenever accept(2) returns
> EMFILE or ENFILE. This eliminates the excessive load of keepalive
> requests on the broker (down to say 100s per second instead of millions per
> second), and allows other parts of the system to proceed and (hopefully)
> close fds so that the next accept() attempt may succeed.
>
> It is also quite simple.
>
> Another approach might be to hold some fds in reserve as described in this
> SO post
> <https://stackoverflow.com/questions/47179793/how-to-gracefully-handle-accept-giving-emfile-and-close-the-connection>,
> however that probably isn't going to work out in the flux-broker since it
> is a threaded process, and something else may grab our reserve fds before
> accept() can use them.
>
> Therefore, at this point I'd lean toward something like this:
>
> diff --git a/src/common/librouter/usock.c b/src/common/librouter/usock.c
> index e1d241c49..c94a1113e 100644--- a/src/common/librouter/usock.c+++ b/src/common/librouter/usock.c@@ -467,6 +467,16 @@ static struct usock_conn *server_accept (struct usock_server *server,
>      return conn;
>  }
>  +static void timeout_cb (flux_reactor_t *r,+                        flux_watcher_t *w,+                        int revents,+                        void *arg)+{+    struct usock_server *server = arg;+    flux_watcher_start (server->w);+    flux_watcher_destroy (w);+    fprintf (stderr, "Starting usock server\n");+}
>
>  static void server_cb (flux_reactor_t *r,
>                         flux_watcher_t *w,@@ -478,8 +488,25 @@ static void server_cb (flux_reactor_t *r,
>      if ((revents & FLUX_POLLIN)) {
>          struct usock_conn *conn;
>  -        if (!(conn = server_accept (server, r)))+        if (!(conn = server_accept (server, r))) {+            if ((errno == ENFILE || errno == EMFILE)+                && !getenv ("USOCK_DISABLE_EMFILE_WORKAROUND")) {+                /*  Too many open files. Do not just back to sleep in the+                 *   reactor since we'll wake right back up again. Instead+                 *   disable this callback until after a short pause, giving+                 *   time for fds to be closed hopefully...+                 */+                flux_watcher_t *tw = flux_timer_watcher_create (r,+                                                                0.01,+                                                                0,+                                                                timeout_cb,+                                                                server);+                flux_watcher_start (tw);+                flux_watcher_stop (w);+                fprintf (stderr, "Too many open files. Pausing usock server\n");+            }
>              return;+        }
>          if (zlist_append (server->connections, conn) < 0) {
>              usock_conn_destroy (conn);
>              return;
>
> I left some debug statements in the patch above to satisfy myself that the
> w/a is activated:
>
> 2021-02-14T20:29:41.381041Z broker.err[0]: socketpair: Too many open files
> 2021-02-14T20:29:41.381946Z broker.err[0]: socketpair: Too many open files
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> Starting usock server
> Too many open files. Pausing usock server
> 2021-02-14T20:29:41.478584Z job-exec.err[0]: exec_start_cmd failed: Too many open files
> Starting usock server
> Too many open files. Pausing usock server
> 2021-02-14T20:29:41.488407Z job-exec.err[0]: exec_start_cmd failed: Too many open files
> Starting usock server
>
> With that patch applied, I cannot reproduce the "hang" anymore, unless
> setting USOCK_DISABLE_EMFILE_WORKAROUND=t in the environment of the test.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/3513#issuecomment-778837570>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPWZODSBG77RDAVE7VLLS7AXSZANCNFSM4XSGYLSA>
> .
>

grondo commented:
Ok, I've added the workaround described above and changed the PR title to be more general.

WIP label is removed since both known issues related to `EMFILE` errors are resolved now.

grondo commented:
Oops, I may have pushed an invalid test image to the `fluxrm/testenv:focal` builder while playing with `docker buildx` I'll fix that up momentarily (it might actually take awhile to rebuild images)

grondo commented:
Ok, fixed. 

I pushed what I think is a functional unit test for the usock server EMFILE handling. Hopefully the test isn't brittle, if it starts failing often in CI or elsewhere we can remove it in the future.

chu11 commented on src/common/librouter/test/usock_emfile.c:
```diff
@@ -0,0 +1,350 @@
+ *
+ * Ensure that both clients connected and successfully sent a message
+ *  each, and count the number of times the server exited the reactor
+ *  and ensure that cont is not unreasonable.
```
typo "count"

chu11 commented:
minor comment/question, are the tests specific to how we open/use file descriptors right now?  i.e. if we use more or less in the future, would the tests fail / unintentionally succeed?  Like:

```
#
￼#  job-exec doesn't segfault when broker runs over nofile ulimit
￼#
￼ulimit -n 100
￼ulimit -Hn 100
￼flux start \
￼    sh -c '
￼flux mini submit --cc=1-12 hostname &&
```

I suppose the 100 limit + 12 jobs was achieved through trial and error?  Perhaps just a comment needed to say these could be needed to be tweaked in future.

grondo commented:
Yeah, 12 jobs was just more than enough to definitely trigger failures, and 100 fds is more than enough to get a flux instance started. If we do start using a lot more fds in the future, the test may fail early. I'll add a comment to that effect.

grondo commented:
Thanks, @chu11. I've fixed the two issues you noted (and squashed them).

garlick reviewed (approved):
Nicely done.  Please set MWP when you're ready.

grondo commented:
Thanks!
